src/core/DataFile.cpp

```diff
@@ -328,6 +405,141 @@ bool DataFile::writeFile( const QString& filename )
 
 
 
+bool DataFile::copyResources(const QString& resourcesDir)
+{
+	// List of filenames used so we can append a counter to any
+	// repeating filenames
+	std::list<QString> namesList;
+
+	ResourcesMap::const_iterator it = ELEMENTS_WITH_RESOURCES.begin();
+
+	// Copy resources and manipulate the DataFile to have local paths to them
+	while (it != ELEMENTS_WITH_RESOURCES.end())
+	{
+		QDomNodeList list = elementsByTagName(it->first);
+
+		// Go through all elements with the tagname from our map
+		for (int i = 0; !list.item(i).isNull(); ++i)
+		{
+			QDomElement el = list.item(i).toElement();
+
+			std::vector<QString>::const_iterator res = it->second.begin();
+
+			// Search for attributes that point to resources
+			while (res != it->second.end())
+			{
+				// If the element has that attribute
+				if (el.hasAttribute(*res))
+				{
+					// Get absolute path to resource
+					QString resPath = PathUtil::toAbsolute(el.attribute(*res));
+					// If we are running from CLI (without the project loaded), "local:" base
+					// prefixes aren't converted properly, so we need to convert it ourselves
+					if (PathUtil::baseLookup(resPath) == PathUtil::Base::LocalDir)
+					{
+						resPath = QFileInfo(m_fileName).path() + "/" + resPath.remove(0,
+							PathUtil::basePrefix(PathUtil::Base::LocalDir).length());
+					}
+
+					// Check if we need to add a counter to the filename
+					QString finalFileName = QFileInfo(resPath).fileName();
+					QString extension = resPath.section('.', -1);
+					int repeatedNames = 0;
+					for (QString name : namesList)
+					{
+						if (finalFileName == name)
+						{
+							++repeatedNames;
+						}
+					}
+					// Add the name to the list before modifying it
+					namesList.push_back(finalFileName);
+					if (repeatedNames)
+					{
+						// Remove the extension, add the counter and add the
+						// extension again to get the final file name
+						finalFileName.truncate(finalFileName.lastIndexOf('.'));
+						finalFileName = finalFileName + "-" + QString::number(repeatedNames) + "." + extension;
+					}
+
+					// Final path is our resources dir + the new file name
+					QString finalPath = resourcesDir + "/" + finalFileName;
+
+					// Copy resource file to the resources folder
+					if(!QFile::copy(resPath, finalPath))
+					{
+						qWarning("ERROR: Failed to copy resource");
+						return false;
+					}
+
+					// Update attribute path to point to the bundle file
+					QString newAtt = "local:resources/" + finalFileName;
+					el.setAttribute(*res, newAtt);
+				}
+				++res;
+			}
+		}
+		++it;
+	}
+
+	return true;
+}
+
+
+
+
+bool DataFile::hasLocalPlugins()
+{
+	QDomNodeList list;
+
+	// First check Vestige plugins
+	list = elementsByTagName("vestige");
+
+	for (int i = 0; !list.item(i).isNull(); ++i)
+	{
+		QDomElement el = list.item(i).toElement();
+
+		if (el.attribute("plugin").startsWith("local:"))
+		{
+			return true;
+		}
+	}
+
+	// Then check effect plugins
```

Comment:
Except for Lv2 (which I think uses unique ID for loading) the external instruments can only be loaded through Vestige right? I might need some help to check whether I covered everything, but those were the two places where I found paths to external libraries.

Grepping `local:` strings still leads us to the problem of having to check what they are a path for. But you kind of gave me an idea: What if instead of looking for the local paths for plugin's we just check if there are any local paths for anything other than the `DataFile::ELEMENTS_WITH_RESOURCES` list, and if there are any we don't load the project? That's probably even safer because then instead of blocking some local paths we are blocking **every** local path except the allowed ones.

Reply:
I like your idea, it should be absolutely safe then, and it's consistent with the other uses of `DataFile::ELEMENTS_WITH_RESOURCES`.

Btw 1: It's correct that Lv2 uses unique IDs (or better called URIDs) for referrencing.
Btw 2: Can this function be made const?

Reply:
Sure! I'll work on making that change then and make the method `const` while I'm at it

Reply:
New method to check for unsafe `local:` uses pushed, also changed it to `const` on the last commit!